librz/util/print.c

```diff
@@ -1872,3 +1926,256 @@ RZ_API void rz_print_rowlog_done(RzPrint *print, const char *str) {
 		}
 	}
 }
+
+static RZ_OWN RzAsmToken *asm_token_create(const size_t start, const size_t len, const RzAsmTokenType type, const ut64 val) {
+	rz_return_val_if_fail(len > 0, NULL);
+	RzAsmToken *t = RZ_NEW0(RzAsmToken);
+
+	t->start = start;
+	t->type = type;
+	t->len = len;
+	switch (type) {
+	default:
+		break;
+	case RZ_ASM_TOKEN_NUMBER:
+		t->val.number = val;
+		break;
+	}
+	return t;
+}
+
+static void add_token(RZ_OUT RzAsmTokenString *token_str, const size_t i, const size_t j, const RzAsmTokenType type, const char *token_val, const bool val_is_hex) {
+	RzAsmToken *t = asm_token_create(i, j - i, type, strtoll(token_val, NULL, val_is_hex ? 16 : 10));
+	if (!t) {
+		RZ_LOG_WARN("Failed to create token. Asm strings will be flawed.\n");
+		rz_warn_if_reached();
+		return;
+	}
+	rz_vector_push(token_str->tokens, t);
+}
+
+static void tokenize_asm_custom(const char *str, RZ_BORROW RzAsmTokenString *toks, RZ_NULLABLE void *mapping) {
+	RZ_LOG_ERROR("Not yet implemented.");
+	rz_warn_if_reached();
+}
+
+/**
+ * \brief Parses an asm string genericly. So only numbers, separators, operations and basic x86 registers
+ * will be recognized as tokens.
+ * Mnemonics will be guessed. Anything else will be marked as an unknown token.
+ *
+ * \param str The plain asm string.
+ * \param toks The parsed string as tokens.
+ */
+static void tokenize_asm_generic(const char *str, RZ_BORROW RzAsmTokenString *toks) {
```

Comment:
i'm not sure it is a good idea to have a generic tokenizer, given that each disassembler/architecture has different rules. sometimes `#` is a comment, other times it may be part of an address IIRC. Maybe the tokenization should be done by the disasm plugins?

cc @XVilka @wargio @thestr4ng3r

Reply:
Yes, I think so. We could keep generic tokenizer meanwhile but mark it as DEPRECATED. Once the we finish the transition of every plugin - it can be removed. What do you think?

Reply:
@thestr4ng3r and I discussed this already a bit on Mattermost. The generic tokenizer works with similar rules as the opcode colorizing function does (see: `print.c::rz_print_colorize_opcode()`).

You are of cause correct. The disasm plugin should be able to customize the token build process.
How it does this needs more consideration. Maybe a list with regex patterns and the associated token type?

Reply:
i agree.

Reply:
> @thestr4ng3r and I discussed this already a bit on Mattermost. The generic tokenizer works with similar rules as the opcode colorizing function does (see: `print.c::rz_print_colorize_opcode()`).
> 
> You are of cause correct. The disasm plugin should be able to customize the token build process. How it does this needs more consideration. Maybe a list with regex patterns and the associated token type?

one thing i learned from r2dec/jsdec is that regexes for parsing asm are absolutely a mess.

Reply:
You can checkout Hexagons `librz/asm/p/asm_hexagon.c::get_token_patterns()`. This is the amount of regex patterns currently needed to parse Hexagons asm strings.
The runtime increase is barely existent (+-0.005s) compared to the generic method (going from left to right over the string).

What problems have you run into?

Reply:
This is how x86 is parsed: https://github.com/wargio/r2dec-js/blob/master/libdec/arch/x86.js#L2187